tree-sitter-tips.md

```diff
@@ -0,0 +1,79 @@
+# Tree-sitter tips and tricks
+
+Language development using [Tree-Sitter](https://github.com/tree-sitter/tree-sitter) can be an arduous task. These tips are intended to help bring empiricism to the process by reducing trial and error (or at least introducing some guard rails).
+
+### :bulb: Rely on Language Specs, but not too much.
+There are language specifications (like [this one for Java](https://docs.oracle.com/javase/specs/jls/se9/html/index.html)) that comprehensively describe a language. Specs provide the lexical and syntactic structure of a language in addition to its components (ex., types, variables, declarations, etc.) and how they interconnect. This can serve as a great baseline for writing your grammar. However, mirroring the spec 100% is not always the best approach, especially when its structure does not align with the parse tree you'd like to generate using tree-sitter. In this situation, it's better to deviate from the spec and prioritize generating a good parse tree. One way to do this may be to find an LR parser for the language you're working on ([such as this one for Java](https://github.com/javaparser/javaparser)) to see how they've modeled their tree outputs. Studying tree outputs from other LR parsers can be a good way to guide your own grammar development. 
```

Comment:
Might be worth noting a couple of concrete deviations we’d want to make:

- supporting multiple versions of a language
- producing a simpler tree by parsing a more lenient grammar than the spec allows
- some specs may encode semantics or context-sensitivity that we want to ignore, e.g. a spec for C might state (I’m spec…ulating, it’s been æons since I’ve read the C spec) that `foo.bar` is only valid when `bar` is a field of some `struct` and `foo` is known to have that type, but since those are both semantic rather than syntactic properties, we would want to parse it regardless
  - in some cases that might even imply ambiguity between different syntactic forms which we’d just have to live with by representing both with a single tree!

Reply:
It would also be good to note the kinds of use cases for these parse trees (e.g. syntax highlighting in a text editor like Atom) as that often drives the level of structure and detail needed.